<a href="https://colab.research.google.com/github/Cho-SangHyun/SwipKkun/blob/main/chatgpt/recommandSystem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. 데이터준비
2. 데이터확인
3. HuggingFace embedding(문자열처리)
4. 코사인유사도 계산(추천 알고리즘)
5. chat gpt이용

임의의 상품 데이터(50,000개)를 통해 임베딩처리 후 코사인유사도를 이용하여 추천시스템을 구축하고 chat gpt를 이용하여 챗봇시스템 구현

#데이터 준비

In [1]:
pip install sentence-transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 11.6 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 110.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 82.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 200.1/200.1 kB 27.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 108.7 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125942 sha256=fc3f2f428ae35cb50c1d573de69cba9b49e5e14765657c679c9935e9054f9433
  Stored in directory: /root/.cache/pip/wheels/71/67/06/162a3760c40d74dd40bc855d527008d26341c2b0ecf3e8e11f
Successfully built sentence-transformers


In [2]:
pip install openai

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 70.3/70.3 kB 1.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 73.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.8/158.8 kB 24.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.2/114.2 kB 12.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 264.6/264.6 kB 26.3 MB/s eta 0:00:00


In [3]:
import requests

import pandas as pd
import numpy as np
import copy
import json

from ast import literal_eval

import torch
from sentence_transformers import SentenceTransformer, util
from transformers import AutoTokenizer, AutoModel
from transformers import OwlViTProcessor, OwlViTForObjectDetection
from transformers import pipeline
from transformers import GPT2TokenizerFast
from PIL import Image

import pickle

In [4]:
import matplotlib.pyplot as plt
from typing import List, Tuple, Dict

import sklearn.datasets as datasets
import sklearn.manifold as manifold

In [5]:
import openai
import os
import sys

openai.api_key = "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"

In [6]:
#google drive mount
from google.colab import drive
drive.mount('/gdrive')

Mounted at /gdrive


In [7]:
#zip파일 압축해제
!unzip -uq "/gdrive/MyDrive/Colab Notebooks/archive.zip" -d"/content"

In [8]:
#csv파일 불러오기
import pandas as pd
appData=pd.read_csv('archive/Sales_Product_Combined.csv')

In [9]:
print(appData.shape)

(185950, 9)


#데이터 확인

In [10]:
appData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185950 entries, 0 to 185949
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order_ID          185950 non-null  int64 
 1   Product           185950 non-null  object
 2   Quantity_Ordered  185950 non-null  int64 
 3   Price             185950 non-null  object
 4   Order_Date        185950 non-null  object
 5   Time              185950 non-null  object
 6   Purchase_Address  185950 non-null  object
 7   City              185950 non-null  object
 8   Product_Type      185950 non-null  object
dtypes: int64(2), object(7)
memory usage: 12.8+ MB


In [11]:
appData.head()

Order_ID                     Product  Quantity_Ordered  Price  Order_Date  \
0    176558        USB-C Charging Cable                 2  11.95  19-04-2019   
1    176559  Bose SoundSport Headphones                 1  99.99  07-04-2019   
2    176560                Google Phone                 1    600  12-04-2019   
3    176560            Wired Headphones                 1  11.99  12-04-2019   
4    176561            Wired Headphones                 1  11.99  30-04-2019   

       Time                      Purchase_Address          City Product_Type  
0   8:46 AM          917 1st St, Dallas, TX 75001        Dallas        Cable  
1  10:30 PM     682 Chestnut St, Boston, MA 02215        Boston   Headphones  
2   2:38 PM  669 Spruce St, Los Angeles, CA 90001   Los Angeles        Phone  
3   2:38 PM  669 Spruce St, Los Angeles, CA 90001   Los Angeles   Headphones  
4   9:27 AM     333 8th St, Los Angeles, CA 90001   Los Angeles   Headphones

In [12]:
# 추천시스템에 사용할 열: Order Id, Product, Quantity Ordered	,Price, Order Date, Product Type
recommended_data=appData[["Order_ID", "Product", "Quantity_Ordered"	,"Price", "Order_Date", "Product_Type"]]
recommended_data.head()

Order_ID                     Product  Quantity_Ordered  Price  Order_Date  \
0    176558        USB-C Charging Cable                 2  11.95  19-04-2019   
1    176559  Bose SoundSport Headphones                 1  99.99  07-04-2019   
2    176560                Google Phone                 1    600  12-04-2019   
3    176560            Wired Headphones                 1  11.99  12-04-2019   
4    176561            Wired Headphones                 1  11.99  30-04-2019   

  Product_Type  
0        Cable  
1   Headphones  
2        Phone  
3   Headphones  
4   Headphones

In [13]:
#데이터 널 값 확인
recommended_data.isnull().sum()

Order_ID            0
Product             0
Quantity_Ordered    0
Price               0
Order_Date          0
Product_Type        0
dtype: int64

In [14]:
#추천서비스를 위한 feature 열 새로 만들기
#str->str만 가능

recommended_data['feature'] = recommended_data['Product'] + " / "   + recommended_data['Product_Type']

<ipython-input-14-8a2c1939726e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  recommended_data['feature'] = recommended_data['Product'] + " / "   + recommended_data['Product_Type']


In [15]:
recommended_data.head()

Order_ID                     Product  Quantity_Ordered  Price  Order_Date  \
0    176558        USB-C Charging Cable                 2  11.95  19-04-2019   
1    176559  Bose SoundSport Headphones                 1  99.99  07-04-2019   
2    176560                Google Phone                 1    600  12-04-2019   
3    176560            Wired Headphones                 1  11.99  12-04-2019   
4    176561            Wired Headphones                 1  11.99  30-04-2019   

  Product_Type                                  feature  
0        Cable             USB-C Charging Cable / Cable  
1   Headphones  Bose SoundSport Headphones / Headphones  
2        Phone                     Google Phone / Phone  
3   Headphones            Wired Headphones / Headphones  
4   Headphones            Wired Headphones / Headphones

#HuggingFace embedding

HugginfFace는 자연어 처리를 위한 오픈소스 라이브러리 및 플랫폼을 개발하는 회사

HuggingFace의 embedding은 단어나 문장을 벡터화하여 자연어 처리작업에서 사용할 수 있도록 도와줌

In [43]:
cur_os = sys.platform
model_path = f"D:/github" if cur_os.startswith('win') else None


In [44]:
if cur_os.startswith('win'):
    model = SentenceTransformer(f'{model_path}/distiluse-base-multilingual-cased-v2')    
else:
    model = SentenceTransformer("sentence-transformers/distiluse-base-multilingual-cased-v2")

model

SentenceTransformer(
  (0): Transformer({'max_seq_length': 128, 'do_lower_case': False}) with Transformer model: DistilBertModel 
  (1): Pooling({'word_embedding_dimension': 768, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
  (2): Dense({'in_features': 768, 'out_features': 512, 'bias': True, 'activation_function': 'torch.nn.modules.activation.Tanh'})
)

In [18]:
#테스트 데이터 수가 200,000개로 너무 커서 이 중 30프로만 사용
recommended_data=recommended_data.sample(frac=0.3, random_state=42)

In [19]:
#185,950->55,785
recommended_data.shape

(55785, 7)

In [20]:
recommended_data['hf_embeddings'] = recommended_data['feature'].apply(lambda x : model.encode(x))
recommended_data.shape

(55785, 8)

In [21]:
#임베딩 결과 확인 (hf_embeddings 열)
print(recommended_data.shape)
recommended_data.head()

(55785, 8)


Order_ID                   Product  Quantity_Ordered     Price  \
75632     149286  Apple Airpods Headphones                 1       150   
30255     295674    AAA Batteries (4-pack)                 4      2.99   
86718     232358        Macbook Pro Laptop                 1  1,700.00   
167590    272349    27in 4K Gaming Monitor                 1    389.99   
41197     306164  Lightning Charging Cable                 1     14.95   

        Order_Date Product_Type                                feature  \
75632   01-01-2019   Headphones  Apple Airpods Headphones / Headphones   
30255   28-12-2019    Batteries     AAA Batteries (4-pack) / Batteries   
86718   21-07-2019       Laptop            Macbook Pro Laptop / Laptop   
167590  21-10-2019      Monitor       27in 4K Gaming Monitor / Monitor   
41197   11-12-2019        Cable       Lightning Charging Cable / Cable   

                                            hf_embeddings  
75632   [0.031172398, 0.09112368, -0.0426553, -0.05605...  
30255   [0.004896611, 0.02960857, 0.013349635, -0.0749...  
86718   [0.00371722, 0.00040116088, 0.009377921, -0.03...  
167590  [0.05769804, 0.07092139, 0.036241647, 0.017500...  
41197   [-0.01632682, -0.011536229, -0.025750225, -0.0...

#코사인 유사도 계산

In [22]:
#추천 아이템 개수 결정
top_k = 5

In [23]:
# SentenceTransformer 모델을 사용하여 주어진 쿼리와 데이터의 문장을 비교하여, 코사인 유사도를 계산하고
# 그 결과를 상위 5 개의 유사한 문장을 반환하는 함수
def get_query_sim_top_k(query, model, df, top_k):
    query_encode = model.encode(query)
    cos_scores = util.pytorch_cos_sim(query_encode, df['hf_embeddings'].tolist())[0]
    top_results = torch.topk(cos_scores, k=top_k)
    return top_results

In [24]:
query = "Are there any Headphones?"
top_result = get_query_sim_top_k(query, model, recommended_data, top_k)

/usr/local/lib/python3.9/dist-packages/sentence_transformers/util.py:39: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:245.)
  b = torch.tensor(b)


In [25]:
#코사인 유사도의 범위  -1 ~1
#두 벡터가 완전히 동일-> 1, 완전히 반대 -> -1 또한 아예 관련이 없는 경우 유사도는 0 
top_result

torch.return_types.topk(
values=tensor([0.5122, 0.5122, 0.5122, 0.5122, 0.5122]),
indices=tensor([20, 12, 29, 56,  5]))

In [26]:
recommended_data.iloc[top_result[1].numpy(), :][['Order_ID', 'Product', 'Quantity_Ordered','Price','Order_Date','Product_Type']]

Order_ID           Product  Quantity_Ordered  Price  Order_Date  \
15669     191584  Wired Headphones                 1  11.99  26-04-2019   
22589     240818  Wired Headphones                 2  11.99  02-08-2019   
131442    205121  Wired Headphones                 1  11.99  13-05-2019   
76500     150113  Wired Headphones                 1  11.99  20-01-2019   
70469     144337  Wired Headphones                 1  11.99  03-01-2019   

       Product_Type  
15669    Headphones  
22589    Headphones  
131442   Headphones  
76500    Headphones  
70469    Headphones

In [27]:
#코사인 유사도를 이용한 함수 test
query = "Do you have a apple product?"
top_result = get_query_sim_top_k(query, model, recommended_data, top_k)
recommended_data.iloc[top_result[1].numpy(), :][['Order_ID', 'Product', 'Quantity_Ordered','Price','Order_Date','Product_Type']]

Order_ID Product  Quantity_Ordered Price  Order_Date Product_Type
68598     142554  iPhone                 1   700  27-01-2019       iPhone
137475    279772  iPhone                 1   700  23-11-2019       iPhone
60672     155713  iPhone                 1   700  18-02-2019       iPhone
167712    272470  iPhone                 1   700  05-10-2019       iPhone
89302     234851  iPhone                 1   700  02-07-2019       iPhone

In [28]:
query = "what is the cheapest laptop?"
top_result = get_query_sim_top_k(query, model, recommended_data, top_k)
recommended_data.iloc[top_result[1].numpy(), :][['Order_ID', 'Product', 'Quantity_Ordered','Price','Order_Date','Product_Type']]

Order_ID          Product  Quantity_Ordered   Price  Order_Date  \
163271    268205  ThinkPad Laptop                 1  999.99  27-10-2019   
16168     192059  ThinkPad Laptop                 1  999.99  25-04-2019   
95163     213714  ThinkPad Laptop                 1  999.99  02-06-2019   
138887    281139  ThinkPad Laptop                 1  999.99  22-11-2019   
18542     236917  ThinkPad Laptop                 1  999.99  24-08-2019   

       Product_Type  
163271       Laptop  
16168        Laptop  
95163        Laptop  
138887       Laptop  
18542        Laptop

#Chatgpt 이용

In [29]:
#gpt3.5 turbo 이용
def print_msg(msg):
    completion = openai.ChatCompletion.create(
                    model="gpt-3.5-turbo",
                    messages=msg
                    )
    return completion['choices'][0]['message']['content']

In [30]:
#prompt test
messages = [
    {"role": "system", "content": "You are a helpful assistant who understands the intent of the user's question."},
    {"role": "user", "content": "Which category does the sentence below belong to: 'description', 'recommended'? Show only categories. \n context: tell me about instagram \n A:"}
]


print_msg(messages)

'Category: Description.'

In [31]:
messages = [
    {"role": "system", "content": "You are a helpful assistant who understands the intent of the user's question."},
    {"role": "user", "content": "Which category does the sentence below belong to: 'description', 'recommend'? Show only categories. \n context: Recommend apps similar to Instagram. \n A:"}
]


print_msg(messages)

'recommend'

In [32]:
msg_prompt = {
    'recom' : {
                'system' : "You are a helpful assistant who recommends the best rental item based on user's requirement in online rental shop", 
                'user' : "Write 1 sentence of a simple greeting that starts with 'Of course!' to recommend items to users.", 
              },
    'desc' : {
                'system' : "You are a helpful assistant who kindly answers.", 
                'user' : "Please write a simple greeting starting with 'of course' to explain the item to the user.", 
              },
     'intent' : {
                'system' : "You are a helpful assistant who understands the intent of the user's question",
                'user' : "Which category does the sentence below belong to: 'description', 'recommended', 'search'? Show only categories. \n context:"
                }         

}

In [33]:
msg_prompt

{'recom': {'system': "You are a helpful assistant who recommends the best rental item based on user's requirement in online rental shop",
  'user': "Write 1 sentence of a simple greeting that starts with 'Of course!' to recommend items to users."},
 'desc': {'system': 'You are a helpful assistant who kindly answers.',
  'user': "Please write a simple greeting starting with 'of course' to explain the item to the user."},
 'intent': {'system': "You are a helpful assistant who understands the intent of the user's question online rental shop.",
  'user': "Which category does the sentence below belong to: 'description', 'recommended', 'search'? Show only categories. \n context:"}}

In [34]:
database=[]

def get_chatgpt_msg(msg):
    completion = openai.ChatCompletion.create(
                    model="gpt-3.5-turbo",
                    messages=msg
                    )
    return completion['choices'][0]['message']['content']

In [35]:
def set_prompt(intent, query, msg_prompt_init, model):
    '''prompt 형태를 만들어주는 함수'''
    m = dict()
    # 검색 또는 추천이면
    if ('recom' in intent) or ('search' in intent):
        msg = msg_prompt_init['recom'] # 시스템 메세지를 가지고오고
    # 설명문이면
    elif 'desc' in intent:
        msg = msg_prompt_init['desc'] # 시스템 메세지를 가지고오고
    # intent 파악
    else:
        msg = msg_prompt_init['intent']
        msg['user'] += f' {query} \n A:'
    for k, v in msg.items():
        m['role'], m['content'] = k, v
    return [m]

In [36]:
def user_interact(query, model, msg_prompt_init):
    # 1. 사용자의 의도를 파악
    user_intent = set_prompt('intent', query, msg_prompt_init, None)
    user_intent = get_chatgpt_msg(user_intent).lower()
    print("user_intent : ", user_intent)
    
    # 2. 사용자의 쿼리에 따라 prompt 생성    
    intent_data = set_prompt(user_intent, query, msg_prompt_init, model)
    intent_data_msg = get_chatgpt_msg(intent_data).replace("\n", "").strip()
    print("intent_data_msg : ", intent_data_msg)
    
    # 3-1. 추천 또는 검색이면
    if ('recom' in user_intent) or ('search' in user_intent):
        recom_msg = str()
        # 기존에 메세지가 있으면 쿼리로 대체
        if len(database) > 0 and database[-1]['role'] == 'assistant' and isinstance(database[-1]['content'], dict):
          query = database[-1]['content']['feature']
        # 유사 아이템 가져오기
        #top_result = get_query_sim_top_k(query, model, movies_metadata, top_k=1 if 'recom' in user_intent else 3) # 추천 개수 설정하려면!
        top_result = get_query_sim_top_k(query, model, recommended_data, top_k=3)
        #print("top_result : ", top_result)
        # 검색이면, 자기 자신의 컨텐츠는 제외
        top_index = top_result[1].numpy() if 'recom' in user_intent else top_result[1].numpy()[1:]
        #print("top_index : ", top_index)
        # Product', 'Price', 'Product_Type를 가져와서 출력
        r_set_d = recommended_data.iloc[top_index, :][['Product', 'Price', 'Product_Type']]
        r_set_d = json.loads(r_set_d.to_json(orient="records"))
        for r in r_set_d:
            for _, v in r.items():
                recom_msg += f"{v} \n"
            recom_msg += "\n"
        database.append({'role' : 'assistant', 'content' : f"{intent_data_msg} {str(recom_msg)}"})
        print(f"\n recom data : {intent_data_msg} \n{str(recom_msg)}")
    # 3-2. 설명이면
    elif 'desc' in user_intent:
        # 이전 메세지에 따라서 설명을 가져와야 하기 때문에 이전 메세지 컨텐츠를 가져옴
        top_result = get_query_sim_top_k(database[-1]['content'], model, recommended_data, top_k=1)
        # feature가 상세 설명이라고 가정하고 해당 컬럼의 값(렌탈아이템)을 가져와 출력
        r_set_d = recommended_data.iloc[top_result[1].numpy(), :][['feature']]
        r_set_d = json.loads(r_set_d.to_json(orient="records"))[0]
        #데이터베이스에 상품에 대한 설명(feature)과 역할을 추가해서 이를 기반으로 설명할 수 있도록
        database.append({'role' : 'assistant', 'content' : r_set_d})
        print(f"\n describe : {intent_data_msg} {r_set_d}")

In [37]:
query = "Do you have a apple product"
user_interact(query, model, copy.deepcopy(msg_prompt))


user_intent :  search
intent_data_msg :  "Of course! I would highly recommend trying our best-selling product, it's loved by many of our satisfied customers."

 recom data : "Of course! I would highly recommend trying our best-selling product, it's loved by many of our satisfied customers." 
 iPhone 
700 
iPhone 

iPhone 
700 
iPhone 




In [38]:
query = "I want to rent a Monitor ."
user_interact(query, model, copy.deepcopy(msg_prompt))


user_intent :  search
intent_data_msg :  "Of course! We have a range of products that you might like, let me show you some options."

 recom data : "Of course! We have a range of products that you might like, let me show you some options." 
 20in Monitor 
109.99 
Monitor 

20in Monitor 
109.99 
Monitor 




In [39]:
query = "I'd like to rent a big screen monitor."
user_interact(query, model, copy.deepcopy(msg_prompt))


user_intent :  search
intent_data_msg :  "Of course! I would highly recommend our best-selling product that has received rave reviews from our customers."

 recom data : "Of course! I would highly recommend our best-selling product that has received rave reviews from our customers." 
 34in Ultrawide Monitor 
379.99 
Monitor 

34in Ultrawide Monitor 
379.99 
Monitor 




# 리뷰

- 테스트 데이터셋이 데이터 수는 많지만 (50,000개 사용) 품목 수가 데이터 수에 비에 다양하지 않고 상품에 대한 설명이 부족해서 대답이 단순함 (34in Ultrawide Monitor / 379.99 / Monitor)
- 상품에 대해서 자세한 설명을 해주기 위해서는 데이터셋(임베딩)에 상품설명을 넣어야 함. (ex 고객이 올린 글, 품번 등)
- query = "I want to rent a Monitor .", query = "I'd like to rent a big screen monitor." 를 보면 질문의 의도나 내용에 대해서는 잘 파악하는 것으로 확인됨
